File: MergeSortedArray.java
/**
 * Created by yhf on 15/4/17.
 */

/*
Given two sorted integer arrays A and B, merge B into A as one sorted array.

Note:
You may assume that A has enough space (size that is greater or equal to m + n) to hold additional
elements from B. The number of elements initialized in A and B are m and n respectively.
*/

public class MergeSortedArray {
    public static void merge(int A[], int m, int B[], int n) {
        /*if (n == 0) {
            return;
        }*/
        int i = 0, j = 0;
        while (i < m && j < n) {
            if (A[i] <= B[j]) {
                i++;
            } else if (A[i] > B[j]) {
                insert(A, i, B[j], m++);
                i++;
                j++;
            }
        }
        while (j < n) {
            A[i++] = B[j++];
        }
    }

    public static void insert(int A[], int index, int num, int size) {
        // insert num into A at index,
        // so elements after index and before size all move backward by 1
        for (int i = size; i > index; i--) {
            A[i] = A[i - 1];
        }
        A[index] = num;
    }

    public static void main(String[] args) {
        int[] A = new int[100];
        int[] B = new int[20];
        A[0] = 2;
        A[1] = 3;
        A[2] = 5;
        A[3] = 7;
        B[0] = 1;
        B[1] = 6;
        B[2] = 7;
        int m = 4, n = 3;
        merge(A, m, B, n);
    }
}

File: Leetcode-Python/MergeSortedArray.py
"""
Given two sorted integer arrays nums1 and nums2, merge nums2 into nums1 as one sorted array.

Note:
You may assume that nums1 has enough space (size that is greater or equal to m + n) to hold additional elements from
nums2. The number of elements initialized in nums1 and nums2 are m and n respectively.
"""


class Solution:
    # @param {integer[]} nums1
    # @param {integer} m
    # @param {integer[]} nums2
    # @param {integer} n
    # @return {void} Do not return anything, modify nums1 in-place instead.
    def merge(self, nums1, m, nums2, n):
        i = j = 0
        end = m
        while i < end and j < n:
            if nums1[i] > nums2[j]:
                nums1.insert(i, nums2[j])
                j += 1
                end += 1
            i += 1
        while j < n:
            nums1[end] = nums2[j]
            end += 1
            j += 1
        while len(nums1) > m + n:
            nums1.pop()


if __name__ == '__main__':
    nums1 = [1, 3, 5, 0, 0, 0, 0, 0, 0, 0]
    nums2 = [2, 4, 6, 8]
    Solution().merge(nums1, 0, nums2, len(nums2))
    print nums1
